elliptic-curve/src/sec1.rs

```diff
@@ -0,0 +1,510 @@
+//! SEC1 encoding support.
+//!
+//! Support for the `Elliptic-Curve-Point-to-Octet-String` encoding format as
+//! described in SEC1: Elliptic Curve Cryptography (Version 2.0) section 2.3.3
+//! (see page 10):
+//!
+//! <https://www.secg.org/sec1-v2.pdf>
+
+use crate::ElementBytes;
+use crate::{
+    point::Generator, scalar::NonZeroScalar, weierstrass::Curve, Arithmetic, Error, FromBytes,
+    SecretKey,
+};
+use core::{
+    convert::TryFrom,
+    fmt::{self, Debug},
+    ops::{Add, Mul},
+};
+use generic_array::{
+    typenum::{Unsigned, U1},
+    ArrayLength, GenericArray,
+};
+use subtle::CtOption;
+
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// Size of an untagged point for given elliptic curve.
+pub type UntaggedPointSize<C> = <<C as crate::Curve>::ElementSize as Add>::Output;
+
+/// SEC1 encoded curve point.
+///
+/// This type is an enum over the compressed and uncompressed encodings,
+/// useful for cases where either encoding can be supported, or conversions
+/// between the two forms.
+#[derive(Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub enum EncodedPoint<C>
+where
+    C: Curve,
+    C::ElementSize: Add<U1>,
+    <C::ElementSize as Add>::Output: Add<U1>,
+    CompressedPointSize<C>: ArrayLength<u8>,
+    UncompressedPointSize<C>: ArrayLength<u8>,
+{
+    /// Compressed Weierstrass elliptic curve point
+    Compressed(CompressedPoint<C>),
+
+    /// Uncompressed Weierstrass elliptic curve point
+    Uncompressed(UncompressedPoint<C>),
+}
```

Comment:
Something we could do here which would be a significant simplification is get rid of `CompressedPoint` and `UncompressedPoint` (and `CompressedPointSize` and `UncompressedPointSize`) entirely and have this type be a struct with a buffer which is large enough to contain either a compressed-or-uncompressed point (which can still be disambiguated by the leading tag byte).

I might take a stab at that as part of this PR.